core/common/src/main/java/alluxio/grpc/GrpcManagedChannelPool.java

```diff
@@ -63,120 +62,107 @@ public static GrpcManagedChannelPool INSTANCE() {
+        Configuration.getMs(PropertyKey.NETWORK_CONNECTION_HEALTH_CHECK_TIMEOUT_MS);
   }
 
-  /**
-   * Restarts the pool by restarting the termination scheduler.
-   */
-  public void restart() throws InterruptedException {
-
-    try (LockResource lockExclusive = new LockResource(mLock.writeLock())) {
-      mScheduler.shutdown();
-      // Wait for each channel termination upto configured single channel timeout
-      long singleChannelTimeoutMs =
-          Configuration.getMs(PropertyKey.MASTER_GRPC_CHANNEL_SHUTDOWN_TIMEOUT);
-      long waitTimeMs = mChannels.size() * singleChannelTimeoutMs;
-      mScheduler.awaitTermination(waitTimeMs, TimeUnit.MILLISECONDS);
-      mScheduler.shutdownNow();
+  private void shutdownManagedChannel(ManagedChannel managedChannel) {
+    managedChannel.shutdown();
+    try {
+      managedChannel.awaitTermination(
+          Configuration.getMs(PropertyKey.MASTER_GRPC_CHANNEL_SHUTDOWN_TIMEOUT),
+          TimeUnit.MILLISECONDS);
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      // Allow thread to exit.
+    } finally {
+      managedChannel.shutdownNow();
     }
-
-    startScheduler();
+    Verify.verify(managedChannel.isShutdown());
   }
 
-  private void startScheduler() {
-
-    mScheduler = Executors.newScheduledThreadPool(1,
-        ThreadFactoryUtils.build("grpc-channel-terminator", true));
-    // Channel termination callback will be fired in the same interval as
-    // the channel shutdown timeout.
-    long channelTerminationIntervalMs =
-        Configuration.getMs(PropertyKey.MASTER_GRPC_CHANNEL_SHUTDOWN_TIMEOUT);
-    mScheduler.scheduleAtFixedRate(() -> {
-      destroyInactiveChannels();
-    }, channelTerminationIntervalMs, channelTerminationIntervalMs, TimeUnit.MILLISECONDS);
-  }
-
-  private void destroyInactiveChannels() {
-    int channelCount = 0;
-    int destroyedCount = 0;
-    List<Pair<ChannelKey, ManagedChannelReference>> channelsToDestroy = new ArrayList<>();
-    try (LockResource lockExclusive = new LockResource(mLock.writeLock())) {
-      channelCount = mChannels.size();
-      for (HashMap.Entry<ChannelKey, ManagedChannelReference> channelEntry : mChannels.entrySet()) {
-        ChannelKey channelKey = channelEntry.getKey();
-        ManagedChannelReference channelReference = channelEntry.getValue();
-        if (channelReference.getRefCount() <= 0) {
-          mChannels.remove(channelKey);
-          channelsToDestroy.add(new Pair(channelKey, channelReference));
-        }
-      }
-    }
-
-    // TODO(ggezer) Consider shutting down in parallel.
-    for (Pair<ChannelKey, ManagedChannelReference> channelPair : channelsToDestroy) {
-      ManagedChannel channel = channelPair.getSecond().get();
-      channel.shutdown();
+  private boolean waitForChannelReady(ManagedChannel managedChannel) {
+    ConnectivityState initialState = managedChannel.getState(false);
+    if (initialState != ConnectivityState.READY) {
```

Comment:
Is it more expensive to always call `managedChannel.getState(true)`?

Reply:
Not really. However passing true doesn't effect the value the method returns. I debugged through it and it triggers traffic after saving the current state, and then return this saved state.

Reply:
Ah I see, thanks